drake/solvers/NloptSolver.cpp

```diff
@@ -0,0 +1,269 @@
+
+#include <iostream>
+#include <stdexcept>
+#include <list>
+#include <vector>
+
+#include <nlopt.hpp>
+
+#include "NloptSolver.h"
+#include "Optimization.h"
+
+namespace Drake {
+namespace {
+Eigen::VectorXd MakeEigenVector(const std::vector<double>& x) {
+  Eigen::VectorXd xvec(x.size());
+  for (size_t i = 0; i < x.size(); i++) {
+    xvec[i] = x[i];
+  }
+  return xvec;
+}
+
+TaylorVecXd MakeInputTaylorVec(const Eigen::VectorXd& xvec,
+                               const VariableList& variable_list) {
+  size_t vars = 0;
+  for (const DecisionVariableView& v : variable_list) {
+    vars += v.size();
+  }
+
+  auto tx = initializeAutoDiff(xvec);
+  TaylorVecXd this_x(vars);
+  size_t index = 0;
+  for (const DecisionVariableView& v : variable_list) {
+      this_x.segment(index, v.size()) = tx.segment(v.index(), v.size());
+      index += v.size();
+    }
+  return this_x;
+}
+
+double EvaluateCosts(const std::vector<double>& x,
+                     std::vector<double>& grad,
+                     void* f_data) {
+  const OptimizationProblem* prog =
+      reinterpret_cast<const OptimizationProblem*>(f_data);
+
+  double cost = 0;
+  Eigen::VectorXd xvec = MakeEigenVector(x);
+
+  auto tx = initializeAutoDiff(xvec);
+  TaylorVecXd ty(1);
+  TaylorVecXd this_x(x.size());
+
+  if (!grad.empty()) {
+    grad.assign(grad.size(), 0);
+  }
+
+  for (auto const& binding : prog->getGenericObjectives()) {
+    size_t index = 0;
+    for (const DecisionVariableView& v : binding.getVariableList()) {
+      this_x.segment(index, v.size()) = tx.segment(v.index(), v.size());
+      index += v.size();
+    }
+
+    binding.getConstraint()->eval(this_x, ty);
+
+    // sam.creasey: I think the right thing to do here is to sum all
+    // of the costs if multiple objectives are specified.
```

Comment:
I think I'll want for further feedback from someone with more expertise (or suggestions for new tests to determine the validity of such an approach).  Unlike constraints, NLopt does not allow us to register multiple objective functions.  If this approach is not valid, we'll probably have to limit the use of the NLopt solver to cases where prog->getGenericObjectives.size() == 1.